Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -305,14 +319,37 @@ public async Task<string> WalkTree(string treeNodeKey)
                         this.walkTreeCts.Token.ThrowIfCancellationRequested();
                     }
 
-                    await this.Parameters.Callbacks.BeforeVisitNode(
-                        this.Parameters.SessionId,
-                        current,
-                        await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
-                        this.Parameters.UserContext,
-                        this.Parameters.TreeName,
-                        this.Parameters.RootSessionId,
-                        this.walkTreeCts.Token).ConfigureAwait(false);
+                    TreeNodeContext treeNodeContext_BeforeVisitNode = null;
+                    if (this.Parameters.CallbacksV2 != null)
+                    {
+                        // If applicable, use the new CallbacksV2 with ITreeWalkerCallbacksV2
+                        treeNodeContext_BeforeVisitNode = new TreeNodeContext(
+                            this.Parameters.SessionId,
+                            current,
+                            await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                            this.Parameters.UserContext,
+                            this.walkTreeCts.Token,
+                            this.Parameters.TreeName,
+                            this.Parameters.RootSessionId,
+                            currentNodeSkipActionContext:null
+                        );
+
+                        // Follow the previous Callbacks with ITreeWalkerCallbacks.
+                        await this.Parameters.CallbacksV2.BeforeVisitNode(treeNodeContext_BeforeVisitNode).ConfigureAwait(false);
+                        this.currentNodeSkipActionContext = treeNodeContext_BeforeVisitNode?.CurrentNodeSkipActionContext;
```

Comment:
>treeNodeContext_BeforeVisitNode? [](start = 60, length = 32)

Are you null checking here in case BeforeVisitNode sets TreeNodeContext to null for some reason? This may be overly safe, but I'm okay with it.

Reply:
good point, removed the "?"

---
In reply to: [505822703](https://github.com/microsoft/Forge/pull/32#discussion_r505822703) [](ancestors = 505822703)